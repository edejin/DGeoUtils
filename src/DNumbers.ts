import {DEGREE_TO_PI, DOUBLE_PI_IN_DEGREE, PI_IN_DEGREE, PI_TO_DEGREE} from './DPoint';

const delta = 0.001;

export class DNumbers {
  static like(v: number, s: number, d: number = delta): boolean {
    return Math.abs(v - s) < d;
  }

  static likeZero(v: number): boolean {
    return DNumbers.like(v, 0);
  }

  static like2PI(v: number): boolean {
    return DNumbers.like(DNumbers.rad2Deg(v), DOUBLE_PI_IN_DEGREE);
  }

  static likePI(v: number): boolean {
    return DNumbers.like(DNumbers.rad2Deg(v), PI_IN_DEGREE);
  }

  static rad2Deg(v: number): number {
    return v * DEGREE_TO_PI;
  }

  static deg2Rad(v: number): number {
    return v * PI_TO_DEGREE;
  }
}
