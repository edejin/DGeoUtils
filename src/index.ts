export * from './DCircle';
export * from './DLine';
export * from './DNumbers';
export * from './DPoint';
export * from './DPolygon';
export * from './FastSearch';
export * from './TraceMatrix';
