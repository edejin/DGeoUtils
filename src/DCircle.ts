import {DPoint, EARTH_RADIUS_IN_METERS} from './DPoint';
import {DPolygon} from './DPolygon';
import {DNumbers} from './DNumbers';
import {checkFunction} from './utils';

// eslint-disable-next-line padded-blocks
export class DCircle {

  /**
   * @param [center=(0,0)]
   * @param [r=0]
   */
  // eslint-disable-next-line no-useless-constructor,no-empty-function
  constructor(public center: DPoint = DPoint.zero(), public r: number = 0) {}

  toString(): string {
    return `(${this.center.toString()}, ${this.r})`;
  }

  getValue(): {center: DPoint, r: number} {
    return {center: this.center, r: this.r};
  }

  clone(): DCircle {
    return new DCircle(this.center, this.r);
  }

  /**
   * Find intersection points with other circle.
   *
   * ![Example](https://edejin.github.io/DGeoUtils/media/examples/findPoints.png)
   * @param c
   */
  findPoints(c: DCircle): DPoint[] | number {
    if (this.equal(c)) {
      return Infinity;
    }
    const {
      center: {x: x0, y: y0},
      r: r0
    } = this;
    const {
      center: {x: x1, y: y1},
      r: r1
    } = c;

    const r02 = r0 * r0;

    const d = this.center.distance(c.center);
    const a = (r02 - r1 * r1 + d * d) / (2 * d);
    const h = Math.sqrt(r02 - a * a);

    const ad = a / d;
    const dy = y1 - y0;
    const dx = x1 - x0;
    const hd = h / d;

    const x2 = x0 + ad * (x1 - x0);
    const y2 = y0 + ad * (y1 - y0);

    const x31 = x2 + hd * dy;
    const y31 = y2 - hd * dx;
    const x32 = x2 - hd * dy;
    const y32 = y2 + hd * dx;

    const res = [];
    if (!isNaN(x31) && !isNaN(y31)) {
      res.push(new DPoint(x31, y31));
    }
    if (!isNaN(x32) && !isNaN(y32) && !(x31 === x32 && y31 === y32)) {
      res.push(new DPoint(x32, y32));
    }
    return res;
  }

  equal({center, r}: DCircle): boolean {
    return this.center.equal(center) && this.r === r;
  }

  /**
   * Transform circle to polygon
   *
   * ![Example](https://edejin.github.io/DGeoUtils/media/examples/findPolygonInside.png)
   * @param [pointCount=64]
   */
  findPolygonInside(pointCount: number = 64): DPolygon {
    const preAngle = 2 * Math.PI / pointCount;
    const points: DPoint[] = [];
    for (let i = 0; i < pointCount; i++) {
      const angle = preAngle * i;
      const x = this.r * Math.cos(angle) + this.center.x;
      const y = this.r * Math.sin(angle) + this.center.y;
      points.push(new DPoint(x, y));
    }
    return new DPolygon([...points, points[0]]);
  }

  /**
   * Transform circle to polygon on sphere. It would be different for different latitude.
   *
   * @remarks
   * Center should be Lng/Lat.
   *
   * @remarks
   * Radius should be in meters.
   *
   * ![Example](https://edejin.github.io/DGeoUtils/media/examples/findPolygonInsideOnSphere.png)
   * @param [pointCount=64]
   */
  findPolygonInsideOnSphere(pointCount: number = 64): DPolygon {
    checkFunction('findPolygonInsideOnSphere')
      .checkArgument('center')
      .shouldBeDegree(this.center);
    const res = new DPolygon();
    for (let i = 0; i < pointCount; i++) {
      res.push(this.sphereOffset(2 * Math.PI * i / pointCount));
    }
    return res.close();
  }

  private sphereOffset(bearing: number, earthRadius = EARTH_RADIUS_IN_METERS): DPoint {
    const lat1 = DNumbers.deg2Rad(this.center.y);
    const lon1 = DNumbers.deg2Rad(this.center.x);
    const dByR = this.r / earthRadius;
    const lat = Math.asin(Math.sin(lat1) * Math.cos(dByR) +
      Math.cos(lat1) * Math.sin(dByR) * Math.cos(bearing));
    const lon =
      lon1 +
      Math.atan2(
        Math.sin(bearing) * Math.sin(dByR) * Math.cos(lat1),
        Math.cos(dByR) - Math.sin(lat1) * Math.sin(lat)
      );
    return new DPoint(DNumbers.rad2Deg(lon), DNumbers.rad2Deg(lat));
  }
}
