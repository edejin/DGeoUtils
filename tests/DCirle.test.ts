/* eslint-disable max-lines,max-statements,max-lines-per-function */
import {DCircle, DGeo, DPoint} from '../src';
import MockInstance = jest.MockInstance;
import 'jest-canvas-mock';

describe('DCircle', () => {
  test('constructor', () => {
    const t = new DCircle(new DPoint(3, 5), 20);
    expect(t.r).toBe(20);
    expect(t.center.equal(new DPoint(3, 5))).toBe(true);
  });

  test('constructor2', () => {
    const t = new DCircle();
    expect(t.r).toBe(0);
    expect(t.center.equal(DPoint.zero())).toBe(true);
  });

  test('toString', () => {
    const t = new DCircle(new DPoint(3, 5), 20);
    expect(t.toString()).toBe('(3 5, 20)');
  });

  test('getValue', () => {
    const t = new DCircle(new DPoint(3, 5), 20);
    expect(t.getValue()).toEqual({
      center: new DPoint(3, 5),
      r: 20
    });
  });

  test('clone', () => {
    const t1 = new DCircle(new DPoint(3, 5), 20);
    const t2 = t1.clone();
    expect(t1.toString()).toBe('(3 5, 20)');
    expect(t2.toString()).toBe('(3 5, 20)');
    t1.center = new DPoint(4, 5);
    t1.r = 30;
    t2.center = new DPoint(6, 7);
    t2.r = 35;
    expect(t1.toString()).toBe('(4 5, 30)');
    expect(t2.toString()).toBe('(6 7, 35)');
  });

  describe('findPoints', () => {
    test('none', () => {
      expect(new DCircle(new DPoint(0, 0), 10).findPoints(new DCircle(new DPoint(30, 0), 10))).toEqual([]);
    });
    test('one', () => {
      expect(new DCircle(new DPoint(0, 0), 10)
        .findPoints(new DCircle(new DPoint(20, 0), 10))).toEqual([new DPoint(10, 0)]);
    });
    test('two withs same y', () => {
      expect(new DCircle(new DPoint(0, 4), 13)
        .findPoints(new DCircle(new DPoint(20, 4), 13)))
        .toEqual([new DPoint(10, -4.306623862918075), new DPoint(10, 12.306623862918075)]);
    });
    test('two with same x', () => {
      expect(new DCircle(new DPoint(3, 0), 13)
        .findPoints(new DCircle(new DPoint(3, 20), 13)))
        .toEqual([
          new DPoint(11.306623862918075, 10),
          new DPoint(-5.306623862918075, 10)
        ]);
    });
    test('two', () => {
      expect(new DCircle(new DPoint(0, 2), 13)
        .findPoints(new DCircle(new DPoint(20, 1), 13)))
        .toEqual([new DPoint(9.585939154181538, -6.781216916369241), new DPoint(10.414060845818462, 9.78121691636924)]);
    });
    test('all', () => {
      expect(new DCircle(new DPoint(0, 0), 11)
        .findPoints(new DCircle(new DPoint(0, 0), 11))).toBe(Infinity);
    });
  });

  describe('findPolygonInside', () => {
    test('1', () => {
      const t = new DCircle(new DPoint(20, 30), 7).findPolygonInside();
      expect(t.toString()).toBe('(' +
        '27 30, ' +
        '26.966293086705377 30.686119982306924, ' +
        '26.86549696282261 31.3656322541129, ' +
        '26.698582350125463 32.031992740781234, ' +
        '26.467156727579006 32.678784026555626, ' +
        '26.173448850438486 33.29977715778198, ' +
        '25.820287286117818 33.888991631137216, ' +
        '25.411073173539158 34.440752989145516, ' +
        '24.949747468305834 34.94974746830583, ' +
        '24.44075298914552 35.41107317353916, ' +
        '23.888991631137216 35.82028728611782, ' +
        '23.299777157781985 36.17344885043848, ' +
        '22.67878402655563 36.46715672757901, ' +
        '22.031992740781234 36.69858235012546, ' +
        '21.365632254112896 36.86549696282261, ' +
        '20.686119982306924 36.96629308670538, ' +
        '19.999999999999996 37, ' +
        '19.313880017693073 36.96629308670538, ' +
        '18.634367745887097 36.86549696282261, ' +
        '17.96800725921876 36.69858235012546, ' +
        '17.321215973444367 36.467156727579, ' +
        '16.70022284221801 36.17344885043848, ' +
        '16.111008368862777 35.82028728611781, ' +
        '15.559247010854476 35.41107317353915, ' +
        '15.050252531694161 34.94974746830583, ' +
        '14.588926826460835 34.44075298914551, ' +
        '14.179712713882179 33.88899163113721, ' +
        '13.82655114956151 33.299777157781975, ' +
        '13.532843272420989 32.67878402655562, ' +
        '13.301417649874534 32.03199274078123, ' +
        '13.134503037177385 31.365632254112885, ' +
        '13.03370691329462 30.68611998230691, ' +
        '13 29.999999999999986, ' +
        '13.033706913294623 29.31388001769306, ' +
        '13.13450303717739 28.634367745887086, ' +
        '13.301417649874542 27.968007259218748, ' +
        '13.532843272421001 27.321215973444357, ' +
        '13.826551149561524 26.700222842218, ' +
        '14.179712713882195 26.11100836886277, ' +
        '14.588926826460854 25.559247010854467, ' +
        '15.05025253169418 25.050252531694152, ' +
        '15.559247010854499 24.588926826460828, ' +
        '16.1110083688628 24.17971271388217, ' +
        '16.700222842218032 23.826551149561507, ' +
        '17.321215973444385 23.532843272420987, ' +
        '17.968007259218776 23.301417649874534, ' +
        '18.63436774588711 23.134503037177385, ' +
        '19.313880017693084 23.033706913294623, ' +
        '20.000000000000004 23, ' +
        '20.686119982306927 23.033706913294623, ' +
        '21.3656322541129 23.13450303717739, ' +
        '22.031992740781234 23.301417649874537, ' +
        '22.678784026555626 23.53284327242099, ' +
        '23.29977715778198 23.82655114956151, ' +
        '23.88899163113721 24.17971271388218, ' +
        '24.44075298914551 24.588926826460835, ' +
        '24.949747468305823 25.050252531694156, ' +
        '25.411073173539148 25.55924701085447, ' +
        '25.820287286117807 26.11100836886277, ' +
        '26.173448850438476 26.700222842217997, ' +
        '26.467156727579 27.32121597344435, ' +
        '26.698582350125456 27.968007259218737, ' +
        '26.865496962822608 28.634367745887076, ' +
        '26.966293086705377 29.313880017693045, ' +
        '27 30' +
        ')');
    });
    test('2', () => {
      const t = new DCircle(new DPoint(20, 30), 7).findPolygonInside(8, 1 / 6 * Math.PI, 5 / 6 * Math.PI);
      expect(t.toString()).toBe('(' +
        '26.062177826491073 33.5, ' +
        '21.811733315717646 36.761480784023476, ' +
        '13.937822173508929 33.5' +
        ')');
    });
    test('3', () => {
      const t = new DCircle(DPoint.zero(), 10).findPolygonInside(64, 0, Math.PI / 2);
      expect(t.toString()).toBe('(' +
        '10 0, ' +
        '9.95184726672197 0.980171403295606, ' +
        '9.807852804032304 1.9509032201612824, ' +
        '9.569403357322088 2.902846772544623, ' +
        '9.238795325112868 3.826834323650898, ' +
        '8.819212643483551 4.7139673682599765, ' +
        '8.314696123025453 5.555702330196022, ' +
        '7.73010453362737 6.343932841636455, ' +
        '7.0710678118654755 7.071067811865475, ' +
        '6.343932841636455 7.730104533627369, ' +
        '5.555702330196023 8.314696123025453, ' +
        '4.713967368259978 8.81921264348355, ' +
        '3.8268343236508984 9.238795325112868, ' +
        '2.902846772544623 9.56940335732209, ' +
        '1.950903220161281 9.807852804032304, ' +
        '6.123233995736766e-16 10' +
        ')');
    });
  });

  describe('findPolygonInsideOnSphere', () => {
    test('1', () => {
      const t = new DCircle(new DPoint(0, 0), 7000).findPolygonInsideOnSphere();
      expect(t.toGeoJSONFeature()).toMatchSnapshot();
    });
    test('2', () => {
      const t = new DCircle(new DPoint(45, 45), 7000).findPolygonInsideOnSphere();
      expect(t.toGeoJSONFeature()).toMatchSnapshot();
    });
    test('3', () => {
      const t = new DCircle(new DPoint(45, 45), 7000).findPolygonInsideOnSphere(64, 0, Math.PI / 2);
      expect(t.toGeoJSONFeature()).toMatchSnapshot();
    });
  });

  describe('checkFunction', () => {
    beforeAll(() => {
      DGeo.DEBUG = true;
    });

    afterAll(() => {
      DGeo.DEBUG = false;
    });

    describe('findPolygonInsideOnSphere', () => {
      // eslint-disable-next-line init-declarations
      let spy: MockInstance<any, any>;
      beforeEach(() => {
        spy = jest.spyOn(console, 'warn').mockImplementation();
      });

      afterEach(() => {
        if (spy) {
          spy.mockRestore();
        }
      });

      test('1', () => {
        expect(new DCircle(new DPoint(1, 10), 2).findPolygonInsideOnSphere()
          .toWKT()).toMatchSnapshot();
        expect(spy).toHaveBeenCalledTimes(7812);
      });
      test('2', () => {
        expect(new DCircle(new DPoint(1, 10010), 2).findPolygonInsideOnSphere()
          .toWKT()).toMatchInlineSnapshot('"POLYGON ((' +
          '1 -69.99998203369482, ' +
          '1.0000051488324002 -69.99998212020742, ' +
          '1.0000102480788564 -69.99998237891208, ' +
          '1.000015248630961 -69.99998280731727, ' +
          '1.0000201023307755 -69.99998340129729, ' +
          '1.0000247624346164 -69.99998415513177, ' +
          '1.0000291840632172 -69.9999850615609, ' +
          '1.0000333246339375 -69.9999861118553, ' +
          '1.0000371442708553 -69.99998729590006, ' +
          '1.0000406061887912 -69.99998860229222, ' +
          '1.0000436770475707 -69.99999001845053, ' +
          '1.0000463272731077 -69.99999153073661, ' +
          '1.0000485313422212 -69.99999312458638, ' +
          '1.0000502680284404 -69.99999478465017, ' +
          '1.0000515206064282 -69.9999964949407, ' +
          '1.0000522770130618 -69.99999823898695, ' +
          '1.000052529963609 -69.99999999999277, ' +
          '1.0000522770218907 -70.00000176099876, ' +
          '1.0000515206237472 -70.00000350504544, ' +
          '1.0000502680535832 -70.00000521533669, ' +
          '1.0000485313742222 -70.00000687540144, ' +
          '1.0000463273107365 -70.00000846925238, ' +
          '1.0000436770893817 -70.00000998153979, ' +
          '1.0000406062331777 -70.00001139769957, ' +
          '1.0000371443161113 -70.00001270409324, ' +
          '1.0000333246783242 -70.00001388813949, ' +
          '1.0000291841050284 -70.00001493843534, ' +
          '1.0000247624722454 -70.00001584486581, ' +
          '1.0000201023627762 -70.00001659870148, ' +
          '1.0000152486561038 -70.00001719268245, ' +
          '1.0000102480961754 -70.00001762108838, ' +
          '1.0000051488412292 -70.00001787979345, ' +
          '1 -70.0000179663062, ' +
          '0.9999948511587708 -70.00001787979345, ' +
          '0.9999897519038247 -70.00001762108838, ' +
          '0.9999847513438962 -70.00001719268245, ' +
          '0.9999798976372238 -70.00001659870148, ' +
          '0.9999752375277547 -70.00001584486581, ' +
          '0.9999708158949717 -70.00001493843534, ' +
          '0.9999666753216759 -70.00001388813949, ' +
          '0.9999628556838887 -70.00001270409324, ' +
          '0.9999593937668223 -70.00001139769957, ' +
          '0.9999563229106182 -70.00000998153979, ' +
          '0.9999536726892635 -70.00000846925238, ' +
          '0.9999514686257778 -70.00000687540144, ' +
          '0.9999497319464168 -70.00000521533669, ' +
          '0.9999484793762529 -70.00000350504544, ' +
          '0.9999477229781094 -70.00000176099876, ' +
          '0.9999474700363911 -69.99999999999277, ' +
          '0.9999477229869383 -69.99999823898695, ' +
          '0.9999484793935718 -69.9999964949407, ' +
          '0.9999497319715597 -69.99999478465017, ' +
          '0.9999514686577787 -69.99999312458638, ' +
          '0.9999536727268924 -69.99999153073661, ' +
          '0.9999563229524294 -69.99999001845053, ' +
          '0.9999593938112088 -69.99998860229222, ' +
          '0.9999628557291448 -69.99998729590006, ' +
          '0.9999666753660625 -69.9999861118553, ' +
          '0.9999708159367828 -69.9999850615609, ' +
          '0.9999752375653836 -69.99998415513177, ' +
          '0.9999798976692246 -69.99998340129729, ' +
          '0.9999847513690391 -69.99998280731727, ' +
          '0.9999897519211435 -69.99998237891208, ' +
          '0.9999948511675999 -69.99998212020742, ' +
          '1 -69.99998203369482))"');
        expect(spy).toHaveBeenCalledTimes(7878);
        expect(spy.mock.calls[0][0]).toEqual('"findPolygonInsideOnSphere" -> "center" should be degree!');
      });
    });
  });
});
