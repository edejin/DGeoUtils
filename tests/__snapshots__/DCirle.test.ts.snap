// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DCircle checkFunction findPolygonInsideOnSphere 1 1`] = `"POLYGON ((1 10.000017966305682, 1.0000017881723582 10.000017879793004, 1.0000035591236365 10.000017621088137, 1.0000052957986028 10.000017192682547, 1.0000069814721257 10.000016598702022, 1.000008599910245 10.000015844866912, 1.0000101355265156 10.000014938437067, 1.0000115735321118 10.0000138881419, 1.0000129000782525 10.000012704096331, 1.0000141023895734 10.000011397703368, 1.0000151688871595 10.000009981544288, 1.0000160893000565 10.000008469257484, 1.0000168547641866 10.0000068754071, 1.0000174579077135 10.000005215342792, 1.0000178929220376 10.000003505051883, 1.0000181556177357 10.00000176100541, 1.0000182434649072 9.999999999999503, 1.0000181556175387 9.999998238993602, 1.0000178929216514 9.99999649494716, 1.000017457907153 9.999994784656296, 1.0000168547634734 9.999993124592049, 1.0000160892992176 9.99999153074174, 1.0000151688862275 9.999990018455023, 1.0000141023885842 9.999988602296035, 1.0000129000772437 9.99998729590317, 1.0000115735311224 9.9999861118577, 1.0000101355255837 9.999985061562624, 1.0000085999094064 9.999984155132864, 1.0000069814714123 9.999983401297833, 1.0000052957980423 9.999982807317366, 1.0000035591232503 9.999982378911824, 1.0000017881721615 9.999982120206985, 1 9.999982033694316, 0.9999982118278385 9.999982120206985, 0.9999964408767497 9.999982378911824, 0.9999947042019576 9.999982807317366, 0.9999930185285877 9.999983401297833, 0.9999914000905937 9.999984155132864, 0.9999898644744164 9.999985061562624, 0.9999884264688776 9.9999861118577, 0.9999870999227561 9.99998729590317, 0.9999858976114159 9.999988602296035, 0.9999848311137725 9.999990018455023, 0.9999839107007823 9.99999153074174, 0.9999831452365268 9.999993124592049, 0.999982542092847 9.999994784656296, 0.9999821070783486 9.99999649494716, 0.9999818443824613 9.999998238993602, 0.9999817565350929 9.999999999999503, 0.9999818443822645 10.00000176100541, 0.9999821070779625 10.000003505051883, 0.9999825420922867 10.000005215342792, 0.9999831452358134 10.0000068754071, 0.9999839106999436 10.000008469257484, 0.9999848311128405 10.000009981544288, 0.9999858976104264 10.000011397703368, 0.9999870999217476 10.000012704096331, 0.9999884264678883 10.0000138881419, 0.9999898644734845 10.000014938437067, 0.9999914000897551 10.000015844866912, 0.9999930185278744 10.000016598702022, 0.9999947042013972 10.000017192682547, 0.9999964408763636 10.000017621088137, 0.9999982118276417 10.000017879793004, 1 10.000017966305682))"`;

exports[`DCircle findPolygonInsideOnSphere 1 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        0,
        0.0628820698883665,
      ],
      Array [
        0.006163523119350049,
        0.06257927541374028,
      ],
      Array [
        0.012267688001591696,
        0.061673808076570726,
      ],
      Array [
        0.01825370807395759,
        0.060174388052581226,
      ],
      Array [
        0.024063934585882962,
        0.058095455623841705,
      ],
      Array [
        0.029642411807028073,
        0.05545703210686164,
      ],
      Array [
        0.03493541591782261,
        0.052284527031134564,
      ],
      Array [
        0.03989197240218116,
        0.04860849342556107,
      ],
      Array [
        0.04446434695935995,
        0.04446433356998517,
      ],
      Array [
        0.0486085052072567,
        0.039891958046142637,
      ],
      Array [
        0.05228453675031367,
        0.03493540137204148,
      ],
      Array [
        0.055457039528643975,
        0.02964239792184763,
      ],
      Array [
        0.058095460747733065,
        0.024063922215712328,
      ],
      Array [
        0.0601743911063653,
        0.01825369800697819,
      ],
      Array [
        0.061673809490249136,
        0.01226768089454847,
      ],
      Array [
        0.06257927577582732,
        0.006163519443017494,
      ],
      Array [
        0.0628820698883665,
        -2.4074827616793682e-17,
      ],
      Array [
        0.06257927577582731,
        -0.006163519443017541,
      ],
      Array [
        0.06167380949024912,
        -0.012267680894548516,
      ],
      Array [
        0.060174391106365285,
        -0.01825369800697824,
      ],
      Array [
        0.05809546074773305,
        -0.024063922215712366,
      ],
      Array [
        0.055457039528643955,
        -0.029642397921847665,
      ],
      Array [
        0.052284536750313644,
        -0.03493540137204152,
      ],
      Array [
        0.048608505207256666,
        -0.03989195804614269,
      ],
      Array [
        0.0444643469593599,
        -0.04446433356998522,
      ],
      Array [
        0.0398919724021811,
        -0.04860849342556111,
      ],
      Array [
        0.03493541591782254,
        -0.05228452703113461,
      ],
      Array [
        0.029642411807027987,
        -0.05545703210686168,
      ],
      Array [
        0.02406393458588286,
        -0.05809545562384175,
      ],
      Array [
        0.018253708073957483,
        -0.06017438805258127,
      ],
      Array [
        0.012267688001591582,
        -0.06167380807657075,
      ],
      Array [
        0.006163523119349924,
        -0.06257927541374028,
      ],
      Array [
        -1.3192546405932253e-16,
        -0.0628820698883665,
      ],
      Array [
        -0.006163523119350187,
        -0.06257927541374027,
      ],
      Array [
        -0.012267688001591838,
        -0.061673808076570705,
      ],
      Array [
        -0.018253708073957736,
        -0.06017438805258119,
      ],
      Array [
        -0.024063934585883108,
        -0.05809545562384164,
      ],
      Array [
        -0.029642411807028222,
        -0.055457032106861565,
      ],
      Array [
        -0.03493541591782275,
        -0.05228452703113447,
      ],
      Array [
        -0.03989197240218131,
        -0.04860849342556094,
      ],
      Array [
        -0.04446434695936008,
        -0.04446433356998504,
      ],
      Array [
        -0.04860850520725683,
        -0.03989195804614249,
      ],
      Array [
        -0.052284536750313776,
        -0.03493540137204133,
      ],
      Array [
        -0.055457039528644045,
        -0.029642397921847485,
      ],
      Array [
        -0.05809546074773311,
        -0.024063922215712203,
      ],
      Array [
        -0.06017439110636533,
        -0.01825369800697809,
      ],
      Array [
        -0.061673809490249164,
        -0.012267680894548393,
      ],
      Array [
        -0.06257927577582734,
        -0.006163519443017446,
      ],
      Array [
        -0.0628820698883665,
        4.429923972593381e-17,
      ],
      Array [
        -0.06257927577582731,
        0.006163519443017533,
      ],
      Array [
        -0.061673809490249136,
        0.01226768089454848,
      ],
      Array [
        -0.0601743911063653,
        0.018253698006978174,
      ],
      Array [
        -0.058095460747733085,
        0.024063922215712283,
      ],
      Array [
        -0.05545703952864401,
        0.029642397921847568,
      ],
      Array [
        -0.05228453675031372,
        0.0349354013720414,
      ],
      Array [
        -0.04860850520725678,
        0.03989195804614256,
      ],
      Array [
        -0.04446434695936004,
        0.04446433356998508,
      ],
      Array [
        -0.03989197240218127,
        0.04860849342556097,
      ],
      Array [
        -0.034935415917822746,
        0.052284527031134474,
      ],
      Array [
        -0.02964241180702824,
        0.05545703210686155,
      ],
      Array [
        -0.024063934585883153,
        0.05809545562384163,
      ],
      Array [
        -0.018253708073957813,
        0.060174388052581164,
      ],
      Array [
        -0.012267688001591944,
        0.06167380807657068,
      ],
      Array [
        -0.006163523119350321,
        0.06257927541374025,
      ],
      Array [
        -1.540167130630379e-17,
        0.0628820698883665,
      ],
    ],
    "type": "LineString",
  },
  "properties": Object {},
  "type": "Feature",
}
`;

exports[`DCircle findPolygonInsideOnSphere 2 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        Array [
          45,
          45.06288206988835,
        ],
        Array [
          45.00872606869534,
          45.06257894353454,
        ],
        Array [
          45.01736782541075,
          45.06167249333455,
        ],
        Array [
          45.025841780758846,
          45.06017147729583,
        ],
        Array [
          45.03406608220303,
          45.05809039713026,
        ],
        Array [
          45.0419613117435,
          45.055449356813796,
        ],
        Array [
          45.04945125898458,
          45.05227386657849,
        ],
        Array [
          45.05646366181996,
          45.048594594317734,
        ],
        Array [
          45.06293090734265,
          45.04444706691203,
        ],
        Array [
          45.06879068603822,
          45.03987132448065,
        ],
        Array [
          45.07398659284551,
          45.03491153102864,
        ],
        Array [
          45.07846866926072,
          45.02961554538361,
        ],
        Array [
          45.08219388131031,
          45.02403445669886,
        ],
        Array [
          45.08512652891437,
          45.01822208913455,
        ],
        Array [
          45.08723858289764,
          45.01223448061416,
        ],
        Array [
          45.08850994666772,
          45.00612934078662,
        ],
        Array [
          45.08892864036126,
          44.999965493504405,
        ],
        Array [
          45.08849090604856,
          44.99380230925324,
        ],
        Array [
          45.08720123337581,
          44.987699133039136,
        ],
        Array [
          45.08507230580394,
          44.98171471325452,
        ],
        Array [
          45.08212486836574,
          44.97590663700774,
        ],
        Array [
          45.078387518600735,
          44.97033077731026,
        ],
        Array [
          45.07389642303519,
          44.9650407573761,
        ],
        Array [
          45.06869496224571,
          44.960087437100434,
        ],
        Array [
          45.06283330817577,
          44.95551842655081,
        ],
        Array [
          45.05636793795997,
          44.95137763103001,
        ],
        Array [
          45.049361089049576,
          44.9477048319546,
        ],
        Array [
          45.0418801609206,
          44.944535307443665,
        ],
        Array [
          45.033997069082126,
          44.94189949613038,
        ],
        Array [
          45.02578755748551,
          44.93982270729824,
        ],
        Array [
          45.017330475764226,
          44.93832488000877,
        ],
        Array [
          45.0087070280087,
          44.93742039343123,
        ],
        Array [
          45,
          44.93711793011163,
        ],
        Array [
          44.9912929719913,
          44.93742039343123,
        ],
        Array [
          44.982669524235774,
          44.93832488000877,
        ],
        Array [
          44.97421244251449,
          44.93982270729824,
        ],
        Array [
          44.966002930917874,
          44.94189949613038,
        ],
        Array [
          44.9581198390794,
          44.944535307443665,
        ],
        Array [
          44.950638910950424,
          44.9477048319546,
        ],
        Array [
          44.94363206204003,
          44.95137763103001,
        ],
        Array [
          44.93716669182423,
          44.95551842655081,
        ],
        Array [
          44.93130503775429,
          44.960087437100434,
        ],
        Array [
          44.92610357696481,
          44.9650407573761,
        ],
        Array [
          44.921612481399265,
          44.97033077731026,
        ],
        Array [
          44.91787513163426,
          44.97590663700774,
        ],
        Array [
          44.91492769419606,
          44.98171471325452,
        ],
        Array [
          44.91279876662419,
          44.987699133039136,
        ],
        Array [
          44.91150909395144,
          44.99380230925324,
        ],
        Array [
          44.91107135963874,
          44.999965493504405,
        ],
        Array [
          44.91149005333228,
          45.00612934078662,
        ],
        Array [
          44.91276141710236,
          45.01223448061416,
        ],
        Array [
          44.91487347108563,
          45.01822208913455,
        ],
        Array [
          44.91780611868969,
          45.02403445669886,
        ],
        Array [
          44.92153133073928,
          45.02961554538361,
        ],
        Array [
          44.92601340715449,
          45.03491153102864,
        ],
        Array [
          44.93120931396178,
          45.03987132448065,
        ],
        Array [
          44.93706909265735,
          45.04444706691203,
        ],
        Array [
          44.94353633818004,
          45.048594594317734,
        ],
        Array [
          44.95054874101541,
          45.05227386657849,
        ],
        Array [
          44.9580386882565,
          45.055449356813796,
        ],
        Array [
          44.96593391779697,
          45.05809039713026,
        ],
        Array [
          44.97415821924115,
          45.06017147729583,
        ],
        Array [
          44.98263217458925,
          45.06167249333455,
        ],
        Array [
          44.99127393130466,
          45.06257894353454,
        ],
        Array [
          45,
          45.06288206988835,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "properties": Object {},
  "type": "Feature",
}
`;

exports[`DCircle findPolygonInsideOnSphere 3 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        45,
        45.06288206988835,
      ],
      Array [
        45.00872606869534,
        45.06257894353454,
      ],
      Array [
        45.01736782541075,
        45.06167249333455,
      ],
      Array [
        45.025841780758846,
        45.06017147729583,
      ],
      Array [
        45.03406608220303,
        45.05809039713026,
      ],
      Array [
        45.0419613117435,
        45.055449356813796,
      ],
      Array [
        45.04945125898458,
        45.05227386657849,
      ],
      Array [
        45.05646366181996,
        45.048594594317734,
      ],
      Array [
        45.06293090734265,
        45.04444706691203,
      ],
      Array [
        45.06879068603822,
        45.03987132448065,
      ],
      Array [
        45.07398659284551,
        45.03491153102864,
      ],
      Array [
        45.07846866926072,
        45.02961554538361,
      ],
      Array [
        45.08219388131031,
        45.02403445669886,
      ],
      Array [
        45.08512652891437,
        45.01822208913455,
      ],
      Array [
        45.08723858289764,
        45.01223448061416,
      ],
      Array [
        45.08892864036126,
        44.999965493504405,
      ],
    ],
    "type": "LineString",
  },
  "properties": Object {},
  "type": "Feature",
}
`;
